### ISSUE TITLE
Expand `rustdoc` support to individual modules

### ISSUE THREAD

**Issue by ojeda** [Rust-for-Linux/linux#181]
Currently we only generate documentation for the shared infrastructure in `rust/`, but we will need to expand the capability to the individual modules.

**Comment by tgross35** [comment]
Has it ever been considered to publish documentation to docs.rs/crates.io? This would be pretty nice for conveniently finding the docs, plus some potential module dev benefits down the line if one could add `linux = 6.1`  as a Cargo dependency instead of needing to maintain the entire tree source. 

I'm not positive but I believe you can reexport non-published crates for them to show up in the docs, e.g.

```rust
#[cfg(doc)]
pub extern crate alloc;

#[cfg(doc)]
pub extern crate bindings;
```

Unfortunately the names [`kernel`](https://docs.rs/kernel/latest/kernel/) and [`linux`](https://docs.rs/linux/0.0.1/linux/) are both taken, but neither has been updated in 6 years and neither really appears used - maybe could reach out to the repo owners and ask if they'd give up the name?

**Comment by bjorn3** [comment]
You can read the documentation at https://rust-for-linux.github.io/docs/kernel/ I don't think it makes sense to use to publish to crates.io as you don't have to just match the kernel and rustc version, but literally have to use the exact same libkernel.rmeta that was compiled when compiling the kernel. And you have to use the same patched liballoc. This is not something cargo can do. Only Kbuild can.

**Comment by ojeda** [comment]
Yeah, in addition, the plan is to get them up at https://docs.kernel.org.

**Comment by tgross35** [comment]
Thanks for the links, I've seen the https://rust-for-linux.github.io pages, but wasn't sure if that would reflect what's actually merged into kernel, or just what is on this branch. https://docs.kernel.org makes more sense here, though I still think the version switcher on https://docs.rs would be nice to see 6.1 vs 6.2, etc (maybe this would be possible on the official kernel docs)

For crates.io, I suppose I had a pipe dream of one day shipping the full source tree and providing a sample `build.rs` script that invokes kbuild, so cargo can handle building OOT modules without manually managing the kernel git. But I suppose that's a really far reach.

**Comment by bjorn3** [comment]
> https://docs.kernel.org/ makes more sense here, though I still think the version switcher on https://docs.rs/ would be nice to see 6.1 vs 6.2, etc (maybe this would be possible on the official kernel docs)

You can view older documentation on docs.kernel.org already. For example https://www.kernel.org/doc/html/v5.19/ for v5.19

> For crates.io, I suppose I had a pipe dream of one day shipping the full source tree and providing a sample build.rs script that invokes kbuild, so cargo can handle building OOT modules without manually managing the kernel git. But I suppose that's a really far reach.

If the kernel you have installed is built with rust support, then you can use a `Makefile` like you would for out-of-tree C modules. (see https://www.kernel.org/doc/html/latest/kbuild/modules.html) It isn't easy to extract the rust for linux code out of the linux repo and in any case building a kernel module requires the kernel source for header files and also requires the kconfig file that was used for building.

**Comment by ojeda** [comment]
> For crates.io, I suppose I had a pipe dream of one day shipping the full source tree and providing a sample build.rs script that invokes kbuild, so cargo can handle building OOT modules without manually managing the kernel git. But I suppose that's a really far reach.

As @bjorn3 points out, kernel modules need to be built for the kernel you are running, since there are hundreds of configuration options. On top of that, kernel internal APIs are unstable. Moreover, distros may patch your kernel too.

But even if it were possible, using Cargo wouldn't bring much to the table, and would be one more interface for users to learn to build the modules. What can be done, though, is using Cargo from the kernel build system when needed (in fact, it is something we already used! :).

To add to the link @bjorn3 added, see https://github.com/Rust-for-Linux/rust-out-of-tree-module too for an example of how a kernel external module may look like.
